README.md
# telegram-melib-bot
[![Build Status](https://img.shields.io/travis/hokus15/melib-telegram-bot?logo=travis)](https://travis-ci.com/github/hokus15/melib-telegram-bot) ![GitHub last commit](https://img.shields.io/github/last-commit/hokus15/melib-telegram-bot?logo=github) ![GitHub commit activity](https://img.shields.io/github/commit-activity/m/hokus15/melib-telegram-bot?logo=github)

Bot de telegram para comprobar las estaciones públicas de carga de vehículo eléctrico libres en las Islas Baleares en un radio de una ubicación proporcionada por el usuario.

La información es obtenida a partir de un servicio web (no publicado oficialmente) de [TIB](https://www.tib.org).

La información que muestra es la misma que puedes obtener desde la web oficial [Mapa de puntos de carga](https://www.tib.org/ximelib/public/map.xhtml)

El bot está desarrollado en Python 3 y desplegado en [Heroku](https://heroku.com).

PRs para mejorar la funcionalidad, la estructura o la legibilidad del código son bienvenidas.

Puedes acceder al bot usando este enlace [@Melib_bot](https://t.me/Melib_bot) o desplegar tu propia instancia en Heroku pulsando el siguiente botón:

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy)

**ATENCIÓN**

Antes de desplegar tu propia instancia en Heroku asegúrate de haber creado un bot siguiendo las intrucciones de la página oficial de [Telegram](https://core.telegram.org/bots#3-how-do-i-create-a-bot).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.2.1 (2021-01-10)
### Fix
* Fix admin message when a restricted user access the bot. ([`eab37f8`](https://github.com/hokus15/melib-telegram-bot/commit/eab37f8c0b45268fcc5a3a0834826fe721522db1))

## v1.2.0 (2020-11-25)
### Feature
* Only show reply keyboard when needed ([`06760a3`](https://github.com/hokus15/melib-telegram-bot/commit/06760a3d1d83ece33351115db7690d953b6bb7a7))
* Add lista and libres commands ([`3da581c`](https://github.com/hokus15/melib-telegram-bot/commit/3da581c247c4ecea334e4397f3210866e5174e19))

### Fix
* Changes in texts returned by the bot and better error handling ([`f138dbd`](https://github.com/hokus15/melib-telegram-bot/commit/f138dbd75874dec297104bfeee12e9db75a02982))
* IF any text is sent to the bot, it will look for free chargers ([`b887132`](https://github.com/hokus15/melib-telegram-bot/commit/b887132f15955215c6c34c5b38cd3772d0bb279b))
* Pop onlyAvailable info instead of getting it ([`9a68c98`](https://github.com/hokus15/melib-telegram-bot/commit/9a68c98c360f5770e675543f3b8f68eca8c3892a))
* Remove reply keyboard when not needed ([`e6ef39b`](https://github.com/hokus15/melib-telegram-bot/commit/e6ef39b8f0c41046aa7a14af3b346e2ee205d9bb))
* Conversation tests ([`0a71919`](https://github.com/hokus15/melib-telegram-bot/commit/0a7191933aaca14d4d9843bdc80e95f334b21933))
* Added needed charger statuses ([`a0e1833`](https://github.com/hokus15/melib-telegram-bot/commit/a0e18334fcce034ab7cbf9ff7bf31b47d3f46753))
* Debugging ([`ca4287c`](https://github.com/hokus15/melib-telegram-bot/commit/ca4287c6105d1e7e279137506b0ff9b81310fc70))
* Added command buttons to reply keyboard ([`804fb68`](https://github.com/hokus15/melib-telegram-bot/commit/804fb68885bcec8deb077068a7b8971f39419944))
* Inlcude offline chargers ([`976c8ec`](https://github.com/hokus15/melib-telegram-bot/commit/976c8ecc8d8fae1ad17e3ff0814e556ddab3fe44))
* Fix fallback method ([`7300b41`](https://github.com/hokus15/melib-telegram-bot/commit/7300b411b07dead8a140f2124cdfdba19491a87a))
* Markup errors in text ([`3d7716a`](https://github.com/hokus15/melib-telegram-bot/commit/3d7716ac2c59e9a3d7fb06a3759b18eba43de84a))
* Fix map link ([`53a01ff`](https://github.com/hokus15/melib-telegram-bot/commit/53a01ffcc044af9b86d1c609c0e3498bfd58a3fb))
* Texts corrections ([`40323d0`](https://github.com/hokus15/melib-telegram-bot/commit/40323d0d9abdbece62ba5cc62717fc38bf279139))
* Fix unit tests ([`5cf17fb`](https://github.com/hokus15/melib-telegram-bot/commit/5cf17fbb6cc998dc5651e3fa4c20f0ad0b6646d1))
* Change parse mode to HTML ([`677cd62`](https://github.com/hokus15/melib-telegram-bot/commit/677cd6254058834561ce379598b96cd71b4cdd94))
* Escape markdown ([`7fd0a03`](https://github.com/hokus15/melib-telegram-bot/commit/7fd0a03f0e92eb28ae40bf483f19e2c8cd65940e))
* Fix misspelling ([`3e4cb9a`](https://github.com/hokus15/melib-telegram-bot/commit/3e4cb9abb7852f48ecc5defd6ecf25605b060a04))

## v1.1.0 (2020-11-19)
### Feature
* Add ADMIN_USER as a env variable instead of taking the first user in VALID_USERS as admin ([`3bc5f4b`](https://github.com/hokus15/melib-telegram-bot/commit/3bc5f4b6c74ea74a1a6b93fa0d632774eecc2337))

### Fix
* Changelog in english ([`ad0aa4b`](https://github.com/hokus15/melib-telegram-bot/commit/ad0aa4b4adb4654291426fafecab53628d4867b9))
* Improve readability of welcome message. ([`08a52b2`](https://github.com/hokus15/melib-telegram-bot/commit/08a52b28af5a31d9437ea9db60fd358ee9786abd))
* Improve messages returned to users ([`fedfa30`](https://github.com/hokus15/melib-telegram-bot/commit/fedfa305265636243195eb6577d3a4fb4d2c411b))
* Fix test ([`e5a9230`](https://github.com/hokus15/melib-telegram-bot/commit/e5a92306fe81f8db550dca73e6402899399b3bbe))
* Fix misspelling in some messages ([`f2e999e`](https://github.com/hokus15/melib-telegram-bot/commit/f2e999e3214fd92990853fe85aff9b8542e84814))
* If no ADMIN_USER env variable has been defined it takes the first user from VALID_USERS as the ADMIN_USER ([`526a401`](https://github.com/hokus15/melib-telegram-bot/commit/526a401ed7234cc5954f0a494ade062b2142c50a))
* Remove date from logging ([`d4c0133`](https://github.com/hokus15/melib-telegram-bot/commit/d4c0133f2b3a903f77d8955a2218b999b780daa8))
* Update method and conversation status names ([`2cf8a2f`](https://github.com/hokus15/melib-telegram-bot/commit/2cf8a2f4469dcf87c0831eaf96559790a51a6ad9))
* Add some logs ([`070484b`](https://github.com/hokus15/melib-telegram-bot/commit/070484b418fac009d7a70dc70be944ecaf83b6c0))

## v1.0.0 (2020-11-12)
### Feature
* Remove Google Cloud Functions support ([`bede341`](https://github.com/hokus15/melib-telegram-bot/commit/bede341705e108f21c3901a9e07e7084e1365a9a))

### Breaking
* With this change the bot will not be able to be  deployed anymore as a Google Cloud Function  ([`bede341`](https://github.com/hokus15/melib-telegram-bot/commit/bede341705e108f21c3901a9e07e7084e1365a9a))

## v0.7.0 (2020-11-11)
### Feature
* Documentation update ([`874d76e`](https://github.com/hokus15/melib-telegram-bot/commit/874d76ec58a9966b1b20700cfe9713817ca5e021))
* Add app.json ([`86e0951`](https://github.com/hokus15/melib-telegram-bot/commit/86e0951ee3a6769426704f9362e5cceb043e6ffd))

### Fix
* Add new env variable to app.json ([`bf67f23`](https://github.com/hokus15/melib-telegram-bot/commit/bf67f23a021f36aaf6bc3b5d94118fa6d868ed59))
* Moooooore docuemtation updates ([`7875d52`](https://github.com/hokus15/melib-telegram-bot/commit/7875d526f4fff0e27d5eef7d7e3bd0cd8ba9a8f9))
* Documentation update ([`9e9bef3`](https://github.com/hokus15/melib-telegram-bot/commit/9e9bef3bccf0135522eddb677d674fc928d99393))
* Better description app.son descriptions ([`2092258`](https://github.com/hokus15/melib-telegram-bot/commit/20922580c52f845b9b2f62bf48446ccce11fb53b))
* Skip conversation tests on non windows environments ([`2178072`](https://github.com/hokus15/melib-telegram-bot/commit/21780722381c008bc4f531473677439aa2e7426f))

test_main.py
import main
import telegram
import json
import flask
import unittest
import re
import os
from telegram import Bot
from unittest.mock import patch
from werkzeug.exceptions import Forbidden

# Evita que falle en travis-ci ya que allí no está el fichero .env
if os.path.exists(".env"):
    from dotenv import load_dotenv
    load_dotenv()


app = flask.Flask(__name__)


class MelibTestCase(unittest.TestCase):

    def test_webhook_autheticate_false(self):
        with open("test_text.json") as f:
            with patch.dict('os.environ', {'TELEGRAM_TOKEN': '0000:yyyy',
                                           'WEBHOOK_TOKEN': 'qwertyuiop',
                                           'VALID_USERS': '1234567890'}):
                with app.test_request_context('?token=poiuytrewq', method="POST", data=f):
                    rq = flask.request
                    try:
                        main.webhook(rq)
                    except Exception as ex:
                        self.assertEqual(type(ex), Forbidden)

    def test_webhook_autheticate_true(self):
        with open("test_text.json") as f:
            with patch.dict('os.environ', {'TELEGRAM_TOKEN': '0000:yyyy',
                                           'WEBHOOK_TOKEN': 'qwertyuiop',
                                           'VALID_USERS': '1234567890'}):
                with patch.object(Bot, "send_message", return_value="test message"):
                    with patch.object(Bot, "send_chat_action", return_value="test action"):
                        with app.test_request_context('?token=qwertyuiop', method="POST", data=f):
                            rq = flask.request
                            rs = main.webhook(rq)
                            self.assertEqual("ok", rs)

    def test_autheticate_true(self):
        with open("test_text.json") as f:
            with patch.dict('os.environ', {'TELEGRAM_TOKEN': '0000:yyyy',
                                           'WEBHOOK_TOKEN': 'qwertyuiop',
                                           'VALID_USERS': '1234567890'}):
                with app.test_request_context('?token=qwertyuiop', method="POST", data=f):
                    r = flask.request
                    self.assertTrue(main._autheticate(r))

    def test_autheticate_false(self):
        with open("test_text.json") as f:
            with patch.dict('os.environ', {'TELEGRAM_TOKEN': '0000:yyyy',
                                           'WEBHOOK_TOKEN': 'qwertyuiop',
                                           'VALID_USERS': '1234567890'}):
                with app.test_request_context('?token=poiuytrewq', method="POST", data=f):
                    r = flask.request
                    self.assertFalse(main._autheticate(r))

    def test_free_chargers_ok(self):
        with open("test_devicegroups_response_ok.json") as f:
            with patch("melib.melib.device_groups") as mock_devicegroups:
                mock_devicegroups.return_value = json.loads(f.read())
                location = telegram.Location(2.4916501, 39.3892373)
                response = main._free_chargers(location)
                self.assertEqual(len(response), 5)

    def test_free_chargers_bad_coordinates(self):
        with open("test_devicegroups_response_bad_coordinates.json") as f:
            with patch("melib.melib.device_groups") as mock_devicegroups:
                mock_devicegroups.return_value = json.loads(f.read())
                location = telegram.Location(2.4916501, 39.3892373)
                response = main._free_chargers(location)
                self.assertEqual(len(response), 4)
                self.assertTrue(684 not in response)

    def test_free_chargers_response_no_chargers(self):
        chargers = {}
        radius = 500
        location = telegram.Location(2.654223, 39.575416)
        response = main._free_chargers_response(chargers, radius, location)
        self.assertTrue(response.startswith('Algo muy gordo ha ocurrido'))

    def test_free_chargers_response_one_charger_available_not_in_radius(self):
        chargers = {
            23: 1000
        }
        radius = 500
        location = telegram.Location(2.654223, 39.575416)
        response = main._free_chargers_response(chargers, radius, location)
        self.assertTrue("No he encontrado cargadores disponibles en 500 metros" in response)
        self.assertTrue(re.search(r"Cargador para \*coche .*\* a \*1000\* metros.*", response, re.DOTALL))
        self.assertEqual(response.count("Cargador para *coche "), 1)

    def test_free_chargers_response_one_charger_in_radius(self):
        chargers = {
            23: 400
        }
        radius = 500
        location = telegram.Location(2.654223, 39.575416)
        response = main._free_chargers_response(chargers, radius, location)
        self.assertTrue('He encontrado los siguientes cargadores' in response)
        self.assertTrue(re.search(r"Cargador para \*coche .*\* a \*400\* metros.*", response, re.DOTALL))
        self.assertEqual(response.count("Cargador para *coche"), 1)

    def test_free_chargers_response_multiple_charger_in_radius(self):
        chargers = {
            23: 400,
            523: 342,
            684: 1203,
            690: 500,
            691: 501
        }
        radius = 500
        location = telegram.Location(2.654223, 39.575416)
        response = main._free_chargers_response(chargers, radius, location)
        self.assertTrue('He encontrado los siguientes cargadores' in response)
        self.assertTrue(re.search(r".*Cargador para \*coche .*\* a \*400\* metros.*", response, re.DOTALL))
        self.assertTrue(re.search(r".*Cargador para \*coche .*\* a \*342\* metros.*", response, re.DOTALL))
        self.assertTrue(re.search(r".*Cargador para \*coche .*\* a \*500\* metros.*", response, re.DOTALL))
        self.assertEqual(response.count("Cargador para *coche "), 3)


test/test_main.py
import main
import flask
import unittest
import os
from telegram import Bot
from unittest.mock import patch
from werkzeug.exceptions import Forbidden

# Evita que falle en travis-ci ya que allí no está el fichero .env
if os.path.exists(".env"):
    from dotenv import load_dotenv
    load_dotenv()


app = flask.Flask(__name__)


class MelibTestCase(unittest.TestCase):

    def test_webhook_autheticate_false(self):
        with open("test/test_text.json") as f:
            with patch.dict('os.environ', {'TELEGRAM_TOKEN': '0000:yyyy',
                                           'WEBHOOK_TOKEN': 'qwertyuiop',
                                           'VALID_USERS': '1234567890'}):
                with app.test_request_context('?token=poiuytrewq', method="POST", data=f):
                    rq = flask.request
                    try:
                        main.webhook(rq)
                    except Exception as ex:
                        self.assertEqual(type(ex), Forbidden)

    def test_webhook_autheticate_true(self):
        with open("test/test_text.json") as f:
            with patch.dict('os.environ', {'TELEGRAM_TOKEN': '0000:yyyy',
                                           'WEBHOOK_TOKEN': 'qwertyuiop',
                                           'VALID_USERS': '1234567890'}):
                with patch.object(Bot, "send_message", return_value="test message"):
                    with patch.object(Bot, "send_chat_action", return_value="test action"):
                        with app.test_request_context('?token=qwertyuiop', method="POST", data=f):
                            rq = flask.request
                            rs = main.webhook(rq)
                            self.assertEqual("ok", rs)

    def test_autheticate_true(self):
        with open("test/test_text.json") as f:
            with patch.dict('os.environ', {'TELEGRAM_TOKEN': '0000:yyyy',
                                           'WEBHOOK_TOKEN': 'qwertyuiop',
                                           'VALID_USERS': '1234567890'}):
                with app.test_request_context('?token=qwertyuiop', method="POST", data=f):
                    r = flask.request
                    self.assertTrue(main._autheticate(r))

    def test_autheticate_false(self):
        with open("test/test_text.json") as f:
            with patch.dict('os.environ', {'TELEGRAM_TOKEN': '0000:yyyy',
                                           'WEBHOOK_TOKEN': 'qwertyuiop',
                                           'VALID_USERS': '1234567890'}):
                with app.test_request_context('?token=poiuytrewq', method="POST", data=f):
                    r = flask.request
                    self.assertFalse(main._autheticate(r))


main.py
# main.py
# To deploy use: gcloud functions deploy webhook

# Fichero usado en Google Cloud Functions
import os
import telegram
from telegram.ext import Dispatcher
from flask import abort
from melibbot import melibbot

bot = None
dispatcher = None


def _autheticate(request):
    '''
    Comprueba si el token proporcionado es válido.
    '''
    webhook_token = os.environ['WEBHOOK_TOKEN']
    request_token = request.args.get('token')
    if request_token and request_token == webhook_token:
        return True
    else:
        return False


def webhook(request):
    # print(request.get_json(force=True))
    global bot
    global dispatcher
    if _autheticate(request):
        if request.method == "POST":
            if bot is None:
                bot = telegram.Bot(token=os.environ.get('TELEGRAM_TOKEN'))
                dispatcher = Dispatcher(bot=bot,
                                        update_queue=None,
                                        workers=0,
                                        use_context=True)
                dispatcher.add_error_handler(melibbot.error_callback)
                dispatcher.add_handler(melibbot.get_handler())

            update = telegram.Update.de_json(request.get_json(force=True), bot)
            dispatcher.process_update(update)
        return "ok"
    else:
        # Si no se ha podido verificar el token
        abort(403)


melib/melib.py
import requests
import json


_RESOURCE = "https://ws.consorcidetransports.com/produccio/ximelib-mobile/rest/"

_HEADERS = {
    "content-type": "application/json",
    "accept-encoding": "gzip",
    "cache-control": "no-cache",
}


def device_groups(payload):
    '''
    Devuelve una lista de los cargadores solicitados con su posición, estado e id.

    Se puede filtrar por los siguentes conceptos:
     placeType: Tipo de plaza.
     Posibles valores:
       CAR: Coche
       MOTORCYCLE: Moto

     idComponentType: Tipo Conector
     Se puede obtener una lista de cargadores válidos usando la siguiente
     petición GET : https://ws.consorcidetransports.com/produccio/ximelib-mobile/rest/devicecomponenttypes
     Posibles valores:
       1: Mennekes
       2: Shuko
       3: CSSCombo
       4: CHAdeMo

     chargeType: Tipo carga Posibles valores:
       CHARGE_MODE_2: Lenta
       CHARGE_MODE_3: Semi-rápida
       CHARGE_MODE_4: Rápida

     onlyAvailable: Mostrar sólo disponibles. true/false

    includeOffline:Mostrar no gestionados. true/false

    bounds:No he podido averiguar para que sirve
    '''
    response = requests.request(
        "POST",
        f"{_RESOURCE}devicegroups",
        data=json.dumps(payload),
        headers=_HEADERS)
    return response.json()


def device_groups_by_id(device_group_id):
    '''
    Devuelve el estado del cargador especificado.
    '''
    response = requests.request("GET", f"{_RESOURCE}devicegroups/{device_group_id}", headers=_HEADERS)
    return response.json()


def device_component_types():
    '''
    Devuelte los tipos de cargadores disponibles.
    '''
    response = requests.request("GET", f"{_RESOURCE}devicecomponenttypes", headers=_HEADERS)
    return response.json()


def sponsors():
    '''
    Devuelve los patrocinadores disponibles.
    '''
    response = requests.request("GET", f"{_RESOURCE}sponsors", headers=_HEADERS)
    return response.json()


melibbot_heroku.py
# Fichero usado en heroku
import logging
import os
from telegram.ext import Updater
from melibbot import melibbot


if __name__ == "__main__":
    WEBHOOK_TOKEN = os.environ.get('WEBHOOK_TOKEN')
    HEROKU_APP_NAME = os.environ.get('HEROKU_APP_NAME')
    TELEGRAM_TOKEN = os.environ.get('TELEGRAM_TOKEN')
    # Port is given by Heroku
    PORT = os.environ.get('PORT', 80)

    # Enable logging
    logging.basicConfig(format='%(name)s - %(levelname)s - %(message)s',
                        level=logging.INFO)
    logger = logging.getLogger(__name__)

    # Set up the Updater
    updater = Updater(token=TELEGRAM_TOKEN)
    dp = updater.dispatcher
    dp.add_handler(melibbot.get_handler())
    dp.add_error_handler(melibbot.error_callback)
    # Start the webhook
    updater.start_webhook(listen="0.0.0.0",
                          port=int(PORT),
                          url_path=WEBHOOK_TOKEN)
    updater.bot.setWebhook("https://{}.herokuapp.com/{}".format(HEROKU_APP_NAME, WEBHOOK_TOKEN))
    updater.idle()


melibbot/melibbot.py
import os
import telegram
import json
import html
import traceback
import logging

from geopy import distance
from functools import wraps
from melibbot import melib
from melibbot.version import __version__
from telegram.ext import (MessageHandler, CallbackQueryHandler,
                          ConversationHandler, Filters, CommandHandler)

logger = logging.getLogger(__name__)

valid_users = os.environ.get('VALID_USERS', '').split(';')

admin_user = os.environ.get('ADMIN_USER', valid_users[0])

MAX_CHARGERS = 10

PLACE_TYPE = {
    'CAR': 'coche',
    'MOTORCYCLE': 'moto'
}

STATUS = {
    'AVAILABLE': 'libre',
    'OCCUPIED_PARTIAL': 'parcialmente ocupado',
    'UNAVAILABLE': 'no disponible',
    'OCCUPIED': 'ocupado',
    'UNKNOWN': 'desconocido',
    'OFFLINE': 'no gestionado'
}

HELP_HEADER = 'Prueba a decirme algo, mandarme los comandos /lista, /libres o una ubicación.'

HELP_LOCATION_INSTRUCTIONS = 'ℹ <b>Para enviar una ubicación</b> ℹ \n' \
                             '1. Pulsa sobre el clip (📎) que encontrarás en la ventana de mensaje.\n' \
                             '2. Elige la opción de `Ubicación`.\n' \
                             '3. Desplázate por el mapa hasta la ubicación que quieras.\n' \
                             '4. Elije `Enviar la ubicación seleccionada`.'

HELP_HINT = 'ℹ <b>CONSEJO</b> ℹ\n' \
            'A parte de enviar tu ubicación actual, puedes enviar la ubicación ' \
            'de tu destino o cualquier otra para saber los cargadores que hay cerca.'

HELP_FOOTER = '‼ <b>ATENCIÓN</b> ‼\n' \
              f'Te voy a devolver como máximo {MAX_CHARGERS} cargadores.\n\n' \
              'La distancia la mido en <b>linea recta</b> entre la ubicación enviada ' \
              'y la ubicación del cargador. No tengo en cuenta la ruta ' \
              'ni la altura de ninguno de los dos puntos. Por lo que la ' \
              'distancia para llegar al cargador puede variar dependiendo ' \
              'del camino que sigas hasta él.\n\n' \
              f'<code>v{__version__}</code>'

# Estados de conversación
LOCATION, RADIUS = range(2)


class BotError(Exception):
    pass


def restricted(func):
    """Decorator: Comprueba si el usuario puede ejecutar el comando `func`."""
    @wraps(func)
    def wrapped(update, context, *args, **kwargs):
        user_id = str(update.effective_user.id)
        if user_id not in valid_users:
            logger.warning(f'Acceso restringido al usuario: {user_id} - '
                           f'{update.effective_user.first_name} {update.effective_user.last_name}')
            update.effective_message.reply_text(
                text=f'Hola {update.effective_user.first_name}, por '
                     'ahora, no puedes usar el bot, por favor, proporciona '
                     f'el siguiente número al administrador:\n\n<b>{user_id}</b>\n\n'
                     'Una vez dado de alta:\n'
                     f'{HELP_HEADER}\n\n'
                     f'{HELP_LOCATION_INSTRUCTIONS}',
                parse_mode=telegram.ParseMode.HTML)
            context.bot.send_message(
                chat_id=admin_user,
                text=f'Hola soy {update.effective_user.first_name} {update.effective_user.last_name} '
                f'y mi usuario de Telegram es:\n<b>{user_id}</b>\nPor favor dame de '
                'alta en el sistema para que pueda acceder al bot.',
                parse_mode=telegram.ParseMode.HTML)
            return
        return func(update, context, *args, **kwargs)
    return wrapped


def send_action(action):
    """Decorator: Envia un `action` mientras procesa el comando `func`."""
    def decorator(func):
        @wraps(func)
        def command_func(update, context, *args, **kwargs):
            context.bot.send_chat_action(chat_id=update.effective_message.chat_id, action=action)
            return func(update, context,  *args, **kwargs)
        return command_func
    return decorator


def error_callback(update, context):
    tb_list = traceback.format_exception(None, context.error, context.error.__traceback__)
    tb = ''.join(tb_list)
    logger.error('Excepción lanzada cuando se procesaba una actualización: {}\n{}'.format(context.error, tb))
    message = (
        'Excepción: {}\n'
        'Mensaje: {}\n'
        '<pre>update = {}</pre>\n\n'
        '<pre>context.chat_data = {}</pre>\n\n'
        '<pre>context.user_data = {}</pre>\n\n'
        '<pre>{}</pre>'
    ).format(
        html.escape(str(type(context.error))),
        html.escape(str(context.error)),
        html.escape(json.dumps(update.to_dict(), indent=2, ensure_ascii=False)),
        html.escape(str(context.chat_data)),
        html.escape(str(context.user_data)),
        html.escape(tb),
    )
    # Corta el mensaje si es más largo de lo permitido
    if len(message) > 4096:
        message = message[:4089] + '</pre>'
    # Envía el mensaje de error al administrador
    context.bot.send_message(chat_id=admin_user, text=message, parse_mode=telegram.ParseMode.HTML)
    update.effective_message.reply_text(text=('Ups! Parece que algo no ha salido bien.\n'
                                              'He enviado un mensaje al administrador con '
                                              'detalles del error para que lo revise.'),
                                        parse_mode=telegram.ParseMode.HTML)
    return ConversationHandler.END


@ restricted
@ send_action(telegram.ChatAction.TYPING)
def free_chargers_help(update, context):
    context.chat_data['onlyAvailable'] = True
    return chargers_help(update, context)


@ restricted
@ send_action(telegram.ChatAction.TYPING)
def list_chargers_help(update, context):
    context.chat_data['onlyAvailable'] = False
    return chargers_help(update, context)


@ restricted
@ send_action(telegram.ChatAction.TYPING)
def chargers_help(update, context):
    location_button = telegram.KeyboardButton(text="Enviar mi ubicación actual", request_location=True)
    reply_markup = telegram.ReplyKeyboardMarkup([[location_button]],
                                                one_time_keyboard=True,
                                                resize_keyboard=True)
    update.effective_message.reply_text(
        text=f'⚠ Hola {update.effective_user.first_name}, para poder darte '
        'información de los cargadores que hay cerca de ti, por favor, '
        '<b>envíame tu ubicación</b> usando el botón de abajo.\n\n'
        f'{HELP_HINT}\n\n'
        f'{HELP_LOCATION_INSTRUCTIONS}\n\n'
        f'{HELP_FOOTER}',
        parse_mode=telegram.ParseMode.HTML,
        reply_markup=reply_markup)
    return LOCATION


@ restricted
@ send_action(telegram.ChatAction.TYPING)
def request_radius(update, context):
    context.chat_data['location'] = json.dumps({
        'latitude': str(update.effective_message.location.latitude),
        'longitude': str(update.effective_message.location.longitude)
    })
    message = '¿Qué radio de búsqueda quieres usar?'
    keyboard = [
        [
            telegram.InlineKeyboardButton('500 m', callback_data='500'),
            telegram.InlineKeyboardButton('1 Km', callback_data='1000'),
            telegram.InlineKeyboardButton('2 Km', callback_data='2000')
        ],
        [
            telegram.InlineKeyboardButton('3 Km', callback_data='3000'),
            telegram.InlineKeyboardButton('5 Km', callback_data='5000'),
            telegram.InlineKeyboardButton('7 Km', callback_data='7000'),
        ],
        # [telegram.InlineKeyboardButton("Cargador libre más cercano", callback_data='0')]
    ]
    reply_markup = telegram.InlineKeyboardMarkup(keyboard)
    update.effective_message.reply_text(message, reply_markup=reply_markup)
    return RADIUS


@ restricted
@ send_action(telegram.ChatAction.TYPING)
def search_chargers(update, context):
    radius = int(update.callback_query.data)
    logger.info('Solicitada búsqueda de cargadores en un radio de {} metros'.format(radius))
    update.callback_query.answer()
    if radius > 0:
        message = ''
        try:
            chat_location = json.loads(context.chat_data['location'])
            location = telegram.Location(float(chat_location['longitude']), float(chat_location['latitude']))
            available_chargers = chargers_list(location, context.chat_data.pop('onlyAvailable', True))
            # Si hay estaciones de carga dentro del radio
            if len(available_chargers) > 0:
                message = chargers_response(available_chargers, radius, location)
            # Si no se han encontrado estaciones de carga dentro del radio
            else:
                message = f'💩 ¡Vaya! No he encontrado ningún cargador en {radius} metros, '
                'comparte otra ubicación y vuelve a probar.'
            context.bot.send_message(chat_id=update.effective_user.id,
                                     parse_mode=telegram.ParseMode.HTML,
                                     disable_web_page_preview=False,
                                     text=message,
                                     reply_markup=telegram.ReplyKeyboardRemove())
        except KeyError as err:
            raise BotError('Error buscando cargadores: {}\ncontext.chat_data: {}.'
                           .format(err, context.chat_data))

    # Si quiere buscar la estación más cercana
    else:
        update.callback_query.edit_message_text(f'Vale {update.callback_query.from_user.first_name}, '
                                                'tú mandas, estación libre más cercana.')

    return ConversationHandler.END


def chargers_list(location, onlyAvailable=True):
    '''
    Devuelve un dict de los cargadores libres o parcialmente ocupados junto
    con la distancia en metros a la ubicación proporcionada.

    La clave del dict es el identificador del cargador y el valor es la distancia
    en metros desde la ubicación proporcionada.
    Ejemplo:
    {
        23: 400,
        523: 342,
        684: 1203,
        690: 500,
        691: 501
    }
    '''

    payload = {
        "bounds": None,
        "idComponentType": None,
        "onlyAvailable": onlyAvailable,
        "includeOffline": True,
        "chargeType": None,
        "placeType": None
    }

    all_chargers_json_data = melib.device_groups(payload)
    latlong = (location.latitude, location.longitude)
    chargers = {}
    for value in all_chargers_json_data:
        charger_id = value['id']
        charger_location = (float(value['lat']), float(value['lng']))
        try:
            # Calcula la distancia en metros desde la estación de carga
            # a la ubicación pasada por el usuario
            dist = distance.distance(latlong, charger_location).meters
            chargers[charger_id] = dist
        except ValueError:
            logger.error(f'Coordenadas {charger_location} incorrectas para el cargador {charger_id}')
    return chargers


def chargers_response(chargers, radius, location):
    '''
    Prepara el texto de respuesta con las estaciones de carga libres dentro del rango.
    '''
    message = ''
    message_charger = ''
    message_header = ''
    message_map_markers = ''
    if len(chargers) > 0:
        # Ordena las estaciones de carga por distancia
        sorted_chargers = sorted(chargers.items(), key=lambda x: x[1])
        # Siempre hay que retornar al menos el cargador más cercano
        closest_charger = sorted_chargers.pop(0)
        closest_charger_data = melib.device_groups_by_id(closest_charger[0])
        closest_charger_distance = closest_charger[1]
        message_charger += f'⚡ 1. {get_charger_text(closest_charger_data, closest_charger_distance)}'
        message_map_markers += f'~{closest_charger_data["lng"]},{closest_charger_data["lat"]},pm2bll1'
        # Si el cargador más cercano está fuera del radio
        if closest_charger_distance > radius:
            message_header = '🤬 No he encontrado cargadores en ' \
                f'{radius} metros, pero el más cercano es:\n\n'
        else:
            message_header = '😁 He encontrado los siguientes cargadores ' \
                f'en {radius} metros:\n\n'
            pos = 2
            for charger in sorted_chargers:
                if charger[1] <= radius and pos <= MAX_CHARGERS:
                    charger_data = melib.device_groups_by_id(charger[0])
                    message_map_markers += f'~{charger_data["lng"]},{charger_data["lat"]},pm2bll{pos}'
                    message_charger += f'⚡ {pos}. {get_charger_text(charger_data, charger[1])}'
                    pos += 1
                else:
                    break
        static_map = 'https://static-maps.yandex.ru/1.x/?lang=es_ES&l=map&ll=' \
            f'{location.longitude},{location.latitude}' \
            f'&size=300,300&pt={location.longitude},{location.latitude},' \
                     f'pm2rdl{message_map_markers}'
        message = f'<a href="{static_map}">🧐</a>{message_header}{message_charger}'
    else:
        logger.error('Parece que no hay ningún cargador disponible')
        message = 'Algo muy gordo ha ocurrido porque no hay ningún cargador disponible en las Baleares'
    return message


def get_charger_text(charger, distance):
    try:
        message = f'Cargador para <b>{PLACE_TYPE[charger["devices"][0]["placeType"]]} ' \
            f'{STATUS[charger["status"]]}</b> a ' \
            f'<b>{distance:0.0f}</b> metros en ' \
            f'<a href="https://www.google.com/maps/place/{charger["lat"]},{charger["lng"]}">' \
            f'<b>{charger["address"]}</b></a>\n'
    except KeyError as err:
        raise BotError('No encuentro la equivalencia de: {}\nError generando los datos del cargador {}.'
                       .format(err, json.dumps(charger)))
    return message


def get_handler():
    conv_handler = ConversationHandler(
        entry_points=[CommandHandler('lista', list_chargers_help),
                      CommandHandler('libres', free_chargers_help),
                      MessageHandler(Filters.location, request_radius),
                      MessageHandler(Filters.text, chargers_help)],
        states={
            LOCATION: [MessageHandler(Filters.location, request_radius)],
            RADIUS: [CallbackQueryHandler(search_chargers)],
        },
        fallbacks=[MessageHandler(Filters.text, chargers_help), MessageHandler(Filters.location, request_radius)]
    )
    return conv_handler
